### PR TITLE
[3.5] Upgrade gRPC-go to 1.58.3

### PR DESCRIPTION
**Pull request by chaochn47**
The last step with gRPC update behavior changes auditing to resolve CVE https://github.com/etcd-io/etcd/issues/16740 in 3.5

This PR backports https://github.com/etcd-io/etcd/pull/14922, https://github.com/etcd-io/etcd/pull/16338, https://github.com/etcd-io/etcd/pull/16587, https://github.com/etcd-io/etcd/pull/16630, https://github.com/etcd-io/etcd/pull/16636 and https://github.com/etcd-io/etcd/pull/16739 to release-3.5. 

Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

**Comment by chaochn47** [comment]
- [ ] https://github.com/grpc/grpc-go/releases/tag/v1.53.0
    - [ ] resolver: remove deprecated field resolver.Target.Endpoint and replace with resolver.Target.Endpoint() (https://github.com/grpc/grpc-go/pull/5852)
    - [ ] solution is to backport https://github.com/etcd-io/etcd/pull/14922
- [ ] https://github.com/grpc/grpc-go/releases/tag/v1.54.0
    - [ ] No behavior changes
- [ ] https://github.com/grpc/grpc-go/releases/tag/v1.55.0
    - [ ] No behavior changes
- [ ] https://github.com/grpc/grpc-go/releases/tag/v1.56.0
    - [ ] No behavior changes
- [ ] https://github.com/grpc/grpc-go/releases/tag/v1.57.0
    - [ ] resolver: remove deprecated Target.Scheme and Target.Authority. Use URL.Scheme and URL.Host instead, respectively (https://github.com/grpc/grpc-go/pull/6363)
    - [ ] etcd does not use those two properties either with internal resolver or etcd backed naming resolver. 
- [ ] https://github.com/grpc/grpc-go/releases/tag/v1.58.0
    - [ ] resolver: add new field Endpoints to State and deprecate Addresses (https://github.com/grpc/grpc-go/pull/6471) Addresses will be deleted in the future. 
    - [ ]  `Addresses` remains and it is not yet removed by gRPC. Not breaking changes.

**Review by jmhbnz — APPROVED** [review]
LGTM - Thanks for all the hard work getting this across the line for 3.5.10 @chaochn47.

**Review by fuweid — APPROVED** [review]
LGTM

**Comment by tjungblu** [comment]
/lgtm

Thanks @chaochn47 :)

**Review by ahrtr — APPROVED** [review]
LGTM

thx @chaochn47

File: client/v3/naming/resolver/resolver.go [hunk 1]
```diff
@@ -17,9 +18,15 @@ type builder struct {
 }
 
 func (b builder) Build(target gresolver.Target, cc gresolver.ClientConn, opts gresolver.BuildOptions) (gresolver.Resolver, error) {
```
Comment (serathius) [inline client/v3/naming/resolver/resolver.go:20]: Is this backported from anywhere? Will try to find it myself, but please separate commits that backport changes as we should not diverge from main branch.

Comment (serathius) [reply client/v3/naming/resolver/resolver.go:20]: Found https://github.com/etcd-io/etcd/pull/14922

**Review by serathius — APPROVED** [review]

**Pull request merged**